// fusion-react/dist/browser.es5.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var React = require('react');
var PropTypes = _interopDefault(require('prop-types'));
var reactIs = require('react-is');
var FusionApp = require('fusion-core');
var FusionApp__default = _interopDefault(FusionApp);
var ReactDOM = _interopDefault(require('react-dom'));

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
var REACT_PREPARE = '@__REACT_PREPARE__@';

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
var prepared = function prepared(sideEffect, opts) {
  if (opts === void 0) {
    opts = {};
  }

  return function (OriginalComponent) {
    opts = Object.assign({
      boundary: false,
      defer: false,
      componentDidMount: true,
      componentWillReceiveProps: false,
      componentDidUpdate: false,
      contextTypes: {},
      forceUpdate: false
    }, opts);
    var prep = {
      prepare: function prepare() {
        return Promise.resolve(sideEffect.apply(void 0, arguments));
      },
      defer: opts.defer
    }; // Disable eslint for deprecated componentWillReceiveProps
    // eslint-disable-next-line react/no-deprecated

    var PreparedComponent =
    /*#__PURE__*/
    function (_React$Component) {
      _inheritsLoose(PreparedComponent, _React$Component);

      // $FlowFixMe
      function PreparedComponent(props, context) {
        var _this;

        _this = _React$Component.call(this, props, context) || this; // $FlowFixMe

        _this[REACT_PREPARE] = prep;
        return _this;
      }

      var _proto = PreparedComponent.prototype;

      _proto.componentDidMount = function componentDidMount() {
        var _this2 = this;

        if (opts.componentDidMount) {
          Promise.resolve(sideEffect(this.props, this.context)).then(function () {
            if (opts.forceUpdate) {
              _this2.forceUpdate();
            }
          });
        }
      }; // $FlowFixMe


      _proto.UNSAFE_componentWillReceiveProps = function UNSAFE_componentWillReceiveProps(nextProps, nextContext) {
        if (opts.componentWillReceiveProps) {
          sideEffect(nextProps, nextContext);
        }
      };

      _proto.componentDidUpdate = function componentDidUpdate() {
        if (opts.componentDidUpdate) {
          sideEffect(this.props, this.context);
        }
      };

      _proto.render = function render() {
        return React.createElement(OriginalComponent, this.props);
      };

      return PreparedComponent;
    }(React.Component);

    var displayName = OriginalComponent.displayName || OriginalComponent.name || '';
    PreparedComponent.contextTypes = opts.contextTypes;
    PreparedComponent.displayName = "PreparedComponent(" + displayName + ")";
    return PreparedComponent;
  };
}; // $FlowFixMe


function isPrepared(CustomComponent) {
  return CustomComponent[REACT_PREPARE] && typeof CustomComponent[REACT_PREPARE].prepare === 'function';
} // $FlowFixMe


function getPrepare(CustomComponent) {
  return CustomComponent[REACT_PREPARE] || {};
}

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
var storeShape = PropTypes.shape({
  dispatch: PropTypes.func.isRequired
}); // $FlowFixMe

var dispatched = function dispatched(prepareUsingDispatch, opts) {
  if (opts === void 0) {
    opts = {};
  }

  return function (OriginalComponent) {
    var prepare = function prepare(props, context) {
      return prepareUsingDispatch(props, context.store.dispatch);
    };

    var contextTypes = Object.assign({}, opts && opts.contextTypes ? opts.contextTypes : {}, {
      store: storeShape
    });
    var preparedOpts = Object.assign({}, opts, {
      contextTypes: contextTypes
    });
    return prepared(prepare, preparedOpts)(OriginalComponent);
  };
};

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
function isReactCompositeComponent(type) {
  if (type != null && type instanceof Function && type.prototype != null && typeof type.prototype.render === 'function') {
    return true;
  }

  return false;
}

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
function isReactFunctionalComponent(type) {
  if (type instanceof Function && (type.prototype == null || !type.prototype.render)) return true;
  return false;
}

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
function renderCompositeElementInstance(instance) {
  var childContext = Object.assign({}, instance.context, instance.getChildContext ? instance.getChildContext() : {});

  if (instance.constructor && instance.constructor.getDerivedStateFromProps) {
    instance.state = _objectSpread({}, instance.state, instance.constructor.getDerivedStateFromProps(instance.props, instance.state));
  } else {
    // see https://github.com/reactjs/react-lifecycles-compat/blob/0a02b805fcf119128d1a9244e71ea7077e2cdcc0/index.js#L114
    if (instance.componentWillMount) {
      instance.componentWillMount();
    } else if (instance.UNSAFE_componentWillMount) {
      instance.UNSAFE_componentWillMount();
    }
  }

  var children = instance.render();
  return [children, childContext];
}

function prepareComponentInstance(instance) {
  if (!isPrepared(instance)) {
    return Promise.resolve({});
  }

  var prepareConfig = getPrepare(instance); // If the component is deferred, skip the prepare step

  if (prepareConfig.defer) {
    return Promise.resolve(prepareConfig);
  } // $FlowFixMe


  return prepareConfig.prepare(instance.props, instance.context).then(function () {
    return prepareConfig;
  });
}

function prepareElement(element, context) {
  if (element === null || _typeof(element) !== 'object') {
    return Promise.resolve([null, context]);
  }

  var type = element.type,
      props = element.props;

  if (reactIs.isContextConsumer(element)) {
    return Promise.resolve([props.children(type._currentValue), context]);
  }

  if (reactIs.isContextProvider(element)) {
    type._context._currentValue = props.value;
    return Promise.resolve([props.children, context]);
  }

  if (typeof type === 'string' || reactIs.isFragment(element) || reactIs.isForwardRef(element)) {
    return Promise.resolve([props.children, context]);
  } else if (isReactFunctionalComponent(type)) {
    return Promise.resolve([type(props, context), context]);
  } else if (isReactCompositeComponent(type)) {
    var CompositeComponent = type;
    var instance = new CompositeComponent(props, context);
    instance.props = props;
    instance.context = context;
    return prepareComponentInstance(instance).then(function (prepareConfig) {
      // Stop traversing if the component is defer or boundary
      if (prepareConfig.defer || prepareConfig.boundary) {
        return Promise.resolve([null, context]);
      }

      return renderCompositeElementInstance(instance);
    });
  } else if (reactIs.isPortal(element)) {
    return Promise.resolve([element.children, context]);
  } else {
    throw new TypeError("Invalid React element type. Must be a string, a function or a subclass of React.Component. " + "This error happens if you write a React Component <Foo> where Foo is undefined. " + "This can happen when mistakenly using a default import instead of a named import or vice-versa, " + "or if you are missing a peerDependency in your package.json and your package manager hoists an older version from an unrelated dependency \n\n" + JSON.stringify(element, null, 2));
  }
}

function _prepare(element, context) {
  return prepareElement(element, context).then(function (_ref) {
    var children = _ref[0],
        childContext = _ref[1];
    return Promise.all(React.Children.toArray(children).map(function (child) {
      return _prepare(child, childContext);
    }));
  });
}

function prepare(element, context) {
  if (context === void 0) {
    context = {};
  }

  context.__IS_PREPARE__ = true;
  return _prepare(element, context).then(function () {
    context.__IS_PREPARE__ = false;
  });
}

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
var CHUNKS_KEY = '__CHUNK_IDS';
var contextTypes = {
  splitComponentLoaders: PropTypes.array.isRequired
};

function withAsyncComponent(_ref) {
  var defer = _ref.defer,
      load = _ref.load,
      LoadingComponent = _ref.LoadingComponent,
      ErrorComponent = _ref.ErrorComponent;
  var AsyncComponent = null;
  var error = null;
  var chunkIds = [];

  function WithAsyncComponent(props) {
    if (error) {
      return React.createElement(ErrorComponent, {
        error: error
      });
    }

    if (!AsyncComponent) {
      return React.createElement(LoadingComponent, null);
    }

    return React.createElement(AsyncComponent, props);
  }

  return prepared(function (props, context) {
    if (AsyncComponent) {
      return Promise.resolve(AsyncComponent);
    }

    var componentPromise;

    try {
      componentPromise = load();
    } catch (e) {
      componentPromise = Promise.reject(e);
    } // $FlowFixMe


    chunkIds = componentPromise[CHUNKS_KEY] || [];

    var loadPromises = [componentPromise].concat(context.splitComponentLoaders.map(function (loader) {
      return loader(chunkIds);
    }));
    return Promise.all(loadPromises).then(function (_ref2) {
      var asyncComponent = _ref2[0];
      // Note: .default is toolchain specific, breaks w/ CommonJS exports
      AsyncComponent = asyncComponent.default;

      if (AsyncComponent === undefined) {
        throw new Error('Bundle does not contain a default export');
      }
    }).catch(function (err) {
      error = err;
      setTimeout(function () {
        throw err;
      }); // log error
    });
  }, {
    defer: defer,
    contextTypes: contextTypes,
    forceUpdate: true
  })(WithAsyncComponent);
}

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
// to visit the minimum number of nodes

var exclude = prepared(Promise.resolve(), {
  componentDidMount: false,
  componentWillReceiveProps: false,
  componentDidUpdate: false,
  defer: true
});

function _inheritsLoose$1(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
var PrepareProvider =
/*#__PURE__*/
function (_React$Component) {
  _inheritsLoose$1(PrepareProvider, _React$Component);

  function PrepareProvider(props, context) {
    var _this;

    _this = _React$Component.call(this, props, context) || this;
    _this.splitComponentLoaders = [];
    _this.markAsCritical = props.markAsCritical;
    return _this;
  }

  var _proto = PrepareProvider.prototype;

  _proto.getChildContext = function getChildContext() {
    return {
      splitComponentLoaders: this.splitComponentLoaders,
      markAsCritical: this.markAsCritical
    };
  };

  _proto.render = function render() {
    return React.Children.only(this.props.children);
  };

  return PrepareProvider;
}(React.Component);

PrepareProvider.childContextTypes = {
  splitComponentLoaders: PropTypes.array.isRequired,
  markAsCritical: PropTypes.func
};

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
var middleware = function middleware(ctx, next) {
  if (process.env.NODE_ENV !== "production") {
    // eslint-disable-next-line no-console
    console.warn('The {middleware} export from fusion-react is deprecated.');
  }

  var markAsCritical = noop;
  ctx.element = React.createElement(PrepareProvider, {
    markAsCritical: markAsCritical
  }, ctx.element);
  return next();
};

function noop() {}

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
var prepareTyped = prepare;
var preparedTyped = prepared;
var splitTyped = withAsyncComponent;
var excludeTyped = exclude;

function _objectSpread$1(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty$1(target, key, source[key]); }); } return target; }

function _defineProperty$1(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _inheritsLoose$2(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
// React.createContext ponyfill
function createContext(value) {
  var _Provider$childContex, _Consumer$contextType;

  if ('createContext' in React) {
    return React.createContext(value);
  }

  var key = "_fusionContextPonyfill" + Math.random();

  var Provider =
  /*#__PURE__*/
  function (_React$Component) {
    _inheritsLoose$2(Provider, _React$Component);

    function Provider() {
      return _React$Component.apply(this, arguments) || this;
    }

    var _proto = Provider.prototype;

    _proto.getChildContext = function getChildContext() {
      var _ref;

      return _ref = {}, _ref[key] = this.props.value || value, _ref;
    };

    _proto.render = function render() {
      return this.props.children;
    };

    return Provider;
  }(React.Component);

  Provider.childContextTypes = (_Provider$childContex = {}, _Provider$childContex[key] = PropTypes.any.isRequired, _Provider$childContex);

  function Consumer(props, context) {
    return props.children(context[key]);
  }

  Consumer.contextTypes = (_Consumer$contextType = {}, _Consumer$contextType[key] = PropTypes.any.isRequired, _Consumer$contextType);
  return {
    Provider: Provider,
    Consumer: Consumer
  };
}

var InjectorContext;

function getServices(app, deps) {
  var services = {};
  Object.entries(deps).forEach(function (_ref2) {
    var name = _ref2[0],
        token = _ref2[1];
    // To be addressed in a future Flow-focued PR.
    // $FlowFixMe
    services[name] = app.getService(token);
  });
  return services;
} // istanbul ignore next


function defaultInject(deps) {
  return {};
}

function defaultMap(services) {
  return services;
}

function registerInjector(app) {
  // Lazily create context for easier testing
  InjectorContext = createContext(defaultInject);

  function inject(deps) {
    return getServices(app, deps);
  }

  function renderProvider(children) {
    return React.createElement(InjectorContext.Provider, {
      value: inject
    }, children);
  }

  var injectorPlugin = FusionApp.createPlugin({
    middleware: function middleware() {
      return function (ctx, next) {
        ctx.element = ctx.element && renderProvider(ctx.element);
        return next();
      };
    }
  });
  app.register(injectorPlugin);
}
function withServices(deps, mapServicesToProps) {
  if (mapServicesToProps === void 0) {
    mapServicesToProps = defaultMap;
  }

  function resolve(inject, props) {
    var services = inject(deps);
    var serviceProps = mapServicesToProps(services);
    return _objectSpread$1({}, serviceProps, props);
  }

  function renderConsumer(Component, props) {
    return React.createElement(InjectorContext.Consumer, null, function (inject) {
      return React.createElement(Component, resolve(inject, props));
    });
  }

  return function (Component) {
    return function WithServices(props) {
      return renderConsumer(Component, props);
    };
  };
}

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */

/* eslint-env node */

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */

/* eslint-env browser */
var clientRender = (function (el) {
  var domElement = document.getElementById('root');

  if (!domElement) {
    throw new Error("Could not find 'root' element");
  }

  return ReactDOM.hydrate ? ReactDOM.hydrate(el, domElement) : ReactDOM.render(el, domElement);
});

function _objectSpread$2(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty$2(target, key, source[key]); }); } return target; }

function _defineProperty$2(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _inheritsLoose$3(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
var Provider = {
  create: function create(name) {
    var _objectSpread2;

    var Provider =
    /*#__PURE__*/
    function (_React$Component) {
      _inheritsLoose$3(Provider, _React$Component);

      function Provider() {
        return _React$Component.apply(this, arguments) || this;
      }

      var _proto = Provider.prototype;

      _proto.getChildContext = function getChildContext() {
        var _ref;

        return _ref = {}, _ref[name] = this.props.provides, _ref;
      };

      _proto.render = function render() {
        return React.Children.only(this.props.children);
      };

      return Provider;
    }(React.Component);

    Provider.childContextTypes = _objectSpread$2({}, Provider.childContextTypes || {}, (_objectSpread2 = {}, _objectSpread2[name] = PropTypes.any.isRequired, _objectSpread2));
    Provider.displayName = name.replace(/^./, function (c) {
      return c.toUpperCase();
    }) + 'Provider';
    return Provider;
  }
};

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
var plugin = {
  create: function create(name, plugin, provider) {
    if (plugin.__plugin__ === undefined) {
      plugin = FusionApp.createPlugin(plugin);
    }

    if (!plugin.__plugin__) {
      throw new Error('Provided plugin does not match FusionPlugin<TDeps, TService>');
    }

    var originalMiddleware = plugin.middleware;
    var ProviderComponent = provider || Provider.create(name);

    plugin.middleware = function (deps, provides) {
      var nextMiddleware = originalMiddleware && originalMiddleware(deps, provides);

      var mw = function mw(ctx, next) {
        if (ctx.element) {
          ctx.element = React.createElement(ProviderComponent, {
            provides: provides,
            ctx: ctx
          }, ctx.element);
        }

        if (nextMiddleware) {
          return nextMiddleware(ctx, next);
        }

        return next();
      };

      return mw;
    };

    return plugin;
  }
};

function _objectSpread$3(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty$3(target, key, source[key]); }); } return target; }

function _defineProperty$3(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _inheritsLoose$4(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
var hoc = {
  create: function create(name, mapProvidesToProps) {
    var mapProvides = mapProvidesToProps ? mapProvidesToProps : function (provides) {
      var _ref;

      return _ref = {}, _ref[name] = provides, _ref;
    };
    return function (Component) {
      var _HOC$contextTypes;

      var HOC =
      /*#__PURE__*/
      function (_React$Component) {
        _inheritsLoose$4(HOC, _React$Component);

        function HOC(props, ctx) {
          var _this;

          _this = _React$Component.call(this, props, ctx) || this;
          _this.provides = ctx[name];
          return _this;
        }

        var _proto = HOC.prototype;

        _proto.render = function render() {
          var props = _objectSpread$3({}, this.props, mapProvides(this.provides));

          return React.createElement(Component, props);
        };

        return HOC;
      }(React.Component);

      var displayName = Component.displayName || Component.name || 'Anonymous';
      HOC.displayName = 'With' + name.replace(/^./, function (c) {
        return c.toUpperCase();
      }) + '(' + displayName + ')';
      HOC.contextTypes = (_HOC$contextTypes = {}, _HOC$contextTypes[name] = PropTypes.any.isRequired, _HOC$contextTypes);
      return HOC;
    };
  }
};

function _inheritsLoose$5(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */

/* eslint-env browser */
var App =
/*#__PURE__*/
function (_FusionApp) {
  _inheritsLoose$5(App, _FusionApp);

  function App(root, render) {
    var _this;

    if (!React.isValidElement(root)) throw new Error('Invalid React element. Ensure your root element is a React.Element (e.g. <Foo />) and not a React.Component (e.g. Foo)');
    var renderer = FusionApp.createPlugin({
      deps: {
        criticalChunkIds: FusionApp.CriticalChunkIdsToken.optional
      },
      provides: function provides() {
        return function (el) {
          return prepareTyped(el).then(function () {
            if (render) {
              return render(el);
            }

            return clientRender(el);
          });
        };
      },
      middleware: function middleware$$1(_ref) {
        var criticalChunkIds = _ref.criticalChunkIds;
        return function (ctx, next) {
          var markAsCritical = noop$1;
          ctx.element = React.createElement(PrepareProvider, {
            markAsCritical: markAsCritical
          }, ctx.element);
          return next();
        };
      }
    });
    _this = _FusionApp.call(this, root, renderer) || this;
    registerInjector(_this);
    return _this;
  }

  return App;
}(FusionApp__default);

function noop$1() {}

exports.default = App;
exports.ProviderPlugin = plugin;
exports.ProvidedHOC = hoc;
exports.Provider = Provider;
exports.withServices = withServices;
exports.dispatched = dispatched;
exports.prepare = prepareTyped;
exports.prepared = preparedTyped;
exports.split = splitTyped;
exports.exclude = excludeTyped;
exports.middleware = middleware;
