// fusion-plugin-react-router/dist/index.es.js
import { createElement, Component, Children } from 'react';
import PropTypes from 'prop-types';
import { Router, Route, BrowserRouter, HashRouter, Link, MemoryRouter, NavLink, Prompt, Switch, matchPath, withRouter } from 'react-router-dom';
import { createPath, parsePath } from 'history';
import { UniversalEventsToken } from 'fusion-plugin-universal-events';
import { createPlugin, createToken, html, memoize } from 'fusion-core';

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
class Status extends Component {
  constructor(props, context) {
    super(props, context);
    const {
      router: {
        staticContext
      }
    } = context;

    if (staticContext && staticContext.setCode) {
      staticContext.setCode(parseInt(this.props.code, 10));
    }
  }

  render() {
    const children = this.props && this.props.children;
    return Array.isArray(children) ? children[0] : children;
  }

}
Status.contextTypes = {
  router: PropTypes.shape({
    staticContext: PropTypes.object
  })
};
const NotFound = props => createElement(Status, {
  code: 404
}, props.children);

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
class Redirect extends Component {
  constructor(props, context) {
    super(props, context);
    if (this.isStatic(context)) this.perform();
  }

  componentDidMount() {
    if (!this.isStatic()) this.perform();
  }

  isStatic(context = this.context) {
    return !!(context && context.router && context.router.staticContext);
  }

  perform() {
    const {
      history,
      staticContext
    } = this.context.router;
    const {
      push,
      to,
      code
    } = this.props;

    if (true && staticContext) {
      staticContext.setCode(parseInt(code, 10));
      staticContext.redirect(to);
      return;
    }

    if (push) {
      history.push(to);
    } else {
      history.replace(to);
    }
  }

  render() {
    return null;
  }

}
Redirect.defaultProps = {
  push: false,
  code: 307
};
Redirect.contextTypes = {
  router: PropTypes.shape({
    history: PropTypes.shape({
      push: PropTypes.func.isRequired,
      replace: PropTypes.func.isRequired
    }).isRequired,
    staticContext: PropTypes.object
  }).isRequired
}; // Sanity type checking

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
const BaseRouter = Router;

/**
 * The public top-level API for a "static" <Router>, so-called because it
 * can't actually change the current location. Instead, it just records
 * location changes in a context object. Useful mainly in testing and
 * server-rendering scenarios.
 */
class ServerRouter extends Component {
  getChildContext() {
    return {
      router: {
        staticContext: this.props.context || {}
      },
      onRoute: routeData => this.props.onRoute && this.props.onRoute(routeData)
    };
  }

  render() {
    const {
      Provider,
      history,
      basename,
      children
    } = this.props;
    if (!Provider) throw new Error('Missing Provider for Server Router');
    return createElement(Provider, {
      basename: basename,
      history: history
    }, children);
  }

}

ServerRouter.defaultProps = {
  basename: '',
  context: {},
  Provider: BaseRouter,
  onRoute: () => {}
};
ServerRouter.childContextTypes = {
  router: () => {},
  onRoute: () => {}
};
const ServerRouterTyped = ServerRouter;

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
const ReactRouterRoute = Route;

const isEmptyChildren = children => Children.count(children) === 0;

function Route$1(props, context) {
  const {
    trackingId,
    component,
    render,
    children
  } = props,
        remainingProps = _objectWithoutProperties(props, ["trackingId", "component", "render", "children"]);

  return createElement(ReactRouterRoute, _extends({}, remainingProps, {
    // eslint-disable-next-line react/no-children-prop
    children: routeProps => {
      const {
        match
      } = routeProps;

      if (match && match.isExact) {
        context.onRoute({
          page: match.path,
          title: trackingId || match.path,
          params: match.params
        });
      }

      if (component) return match ? createElement(component, routeProps) : null;
      if (render) return match ? render(routeProps) : null;
      if (typeof children === 'function') return children(routeProps);
      if (children && !isEmptyChildren(children)) return Children.only(children);
      return null;
    }
  }));
}

Route$1.contextTypes = {
  onRoute: PropTypes.func.isRequired
};
Route$1.displayName = 'FusionRoute';

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
/**
 * Cast each of these imports from react-router-dom to a copied-version of their
 * types.  This is necessary as the libdef defined types will not be accessible to
 * consumers of this package.
 */
const BrowserRouter$1 = BrowserRouter;
const HashRouter$1 = HashRouter;
const Link$1 = Link;
const MemoryRouter$1 = MemoryRouter;
const NavLink$1 = NavLink;
const Prompt$1 = Prompt;
const Switch$1 = Switch;
const matchPath$1 = matchPath;
const withRouter$1 = withRouter;

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
const BaseRouter$1 = Router;

class BrowserRouter$2 extends Component {
  constructor(props, context) {
    super(props, context);
    this.lastTitle = null;
  }

  getChildContext() {
    const {
      __IS_PREPARE__
    } = this.context;
    return {
      onRoute: routeData => {
        if (routeData.title !== this.lastTitle && !__IS_PREPARE__) {
          this.lastTitle = routeData.title;
          this.props.onRoute && this.props.onRoute(routeData);
        }
      }
    };
  }

  render() {
    const {
      Provider,
      history,
      basename
    } = this.props;
    if (!Provider) throw new Error('Missing Provider for Browser Router');
    return createElement(Provider, {
      basename: basename,
      history: history
    }, this.props.children);
  }

}

BrowserRouter$2.defaultProps = {
  onRoute: () => {},
  Provider: BaseRouter$1
};
BrowserRouter$2.propTypes = {
  children: PropTypes.node,
  onRoute: PropTypes.func,
  history: PropTypes.object,
  Provider: PropTypes.any,
  basename: PropTypes.string
};
BrowserRouter$2.contextTypes = {
  __IS_PREPARE__: PropTypes.bool
};
BrowserRouter$2.childContextTypes = {
  onRoute: PropTypes.func.isRequired
};

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const addLeadingSlash = path => path.charAt(0) === '/' ? path : '/' + path;

const addRoutePrefix = (location, prefix) => {
  if (!prefix) return location;

  if (typeof location === 'string') {
    return `${prefix}${addLeadingSlash(location)}`;
  } else {
    return _objectSpread({}, location, {
      pathname: `${prefix}${addLeadingSlash(location.pathname)}`
    });
  }
};
const removeRoutePrefix = (location, prefix) => {
  if (!prefix) return location;
  const pathname = typeof location === 'string' ? location : location.pathname;
  const hasPrefix = (pathname + '/').indexOf(prefix + '/') === 0;
  const unprefixedPathname = pathname.slice(prefix.length);
  const relativePathname = hasPrefix ? unprefixedPathname : pathname;

  if (typeof location === 'string') {
    return relativePathname;
  } else {
    return _objectSpread({}, location, {
      pathname: relativePathname
    });
  }
};

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
const createLocation = (path, prefix) => {
  const unprefixedPath = removeRoutePrefix(path, prefix);
  return parsePath(unprefixedPath);
};

const createPrefixedURL = (location, prefix) => {
  if (typeof location === 'string') {
    return addRoutePrefix(location, prefix);
  } else {
    return createPath(addRoutePrefix(location, prefix));
  }
};

const createURL = (location, prefix) => {
  if (typeof location === 'string') {
    return removeRoutePrefix(location, prefix);
  } else {
    return createPath(removeRoutePrefix(location, prefix));
  }
};

const staticHandler = methodName => () => {
  throw new Error(`You cannot ${methodName} with server side <Router>`);
};

const noop = () => {};

function createServerHistory(basename, context, location) {
  function createHref(location) {
    return createPrefixedURL(location, basename);
  }

  function push(path) {
    context.action = 'PUSH';
    context.location = createLocation(path, basename);
    const url = createURL(path, basename);

    if (typeof url === 'string') {
      context.url = url;
    }
  }

  function replace(path) {
    context.action = 'REPLACE';
    context.location = createLocation(path, basename);
    const url = createURL(path, basename);

    if (typeof url === 'string') {
      context.url = url;
    }
  }

  const history = {
    length: 0,
    createHref,
    action: 'POP',
    location: createLocation(location, basename),
    push,
    replace,
    go: staticHandler('go'),
    goBack: staticHandler('back'),
    goForward: staticHandler('forward'),
    listen: () => noop
  };
  return history;
}

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
const RouterProviderToken = createToken('RouterProvider');
const RouterToken = createToken('Router');
const Router$1 = ServerRouterTyped;
const plugin = createPlugin({
  deps: {
    emitter: UniversalEventsToken.optional,
    Provider: RouterProviderToken.optional
  },
  middleware: ({
    emitter,
    Provider = Router
  }, self) => {
    return async (ctx, next) => {
      const prefix = ctx.prefix || '';

      if (!ctx.element) {
        return next();
      }

      const myAPI = self.from(ctx);

      {
        let pageData = {
          title: ctx.path,
          page: ctx.path
        };
        const context = {
          action: null,
          location: null,
          url: null,
          setCode: code => {
            ctx.status = code;
          },
          redirect: url => {
            const toUrl = addRoutePrefix(url, prefix);

            if (typeof toUrl === 'string') {
              ctx.redirect(toUrl);
            }
          }
        }; // Expose the history object

        const history = createServerHistory(prefix, context, prefix + ctx.url);
        myAPI.history = history;
        ctx.element = createElement(Router$1, {
          history: history,
          Provider: Provider,
          onRoute: d => {
            pageData = d;
          },
          basename: prefix,
          context: context
        }, ctx.element);
        return next().then(() => {
          ctx.template.body.push(html`
              <script id="__ROUTER_DATA__" type="application/json">
                ${JSON.stringify(pageData)}
              </script>
            `);

          if (emitter) {
            const scopedEmitter = emitter.from(ctx);

            const emitTiming = type => timing => {
              scopedEmitter.emit(type, {
                title: pageData.title,
                page: pageData.page,
                status: ctx.status,
                timing
              });
            };

            scopedEmitter.map(payload => {
              if (payload && typeof payload === 'object') {
                payload.__url__ = pageData.title;
              }

              return payload;
            });
            ctx.timing.end.then(timing => {
              emitTiming('pageview:server')(timing);
              ctx.timing.render.then(emitTiming('render:server'));
            });
          }
        });
      }
    };
  },

  provides() {
    return {
      from: memoize(() => {
        const api = {
          history: null
        };
        return api;
      })
    };
  }

});

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
const BrowserRouter$4 = BrowserRouter$1;
const HashRouter$3 = HashRouter$1;
const Link$3 = Link$1;
const matchPath$3 = matchPath$1;
const MemoryRouter$3 = MemoryRouter$1;
const NavLink$3 = NavLink$1;
const Prompt$3 = Prompt$1;
const Route$2 = Route$1;
const Router$2 = ServerRouterTyped;
const Switch$3 = Switch$1;
const withRouter$3 = withRouter$1;
const NotFound$1 = NotFound;
const Redirect$1 = Redirect;
const Status$1 = Status;

export default plugin;
export { BrowserRouter$4 as BrowserRouter, HashRouter$3 as HashRouter, Link$3 as Link, matchPath$3 as matchPath, MemoryRouter$3 as MemoryRouter, NavLink$3 as NavLink, NotFound$1 as NotFound, Prompt$3 as Prompt, Redirect$1 as Redirect, Route$2 as Route, Router$2 as Router, Status$1 as Status, Switch$3 as Switch, withRouter$3 as withRouter, RouterProviderToken, RouterToken };
